### ISSUE TITLE
[trash-cli] support on linux please

### ISSUE THREAD

**Issue by IntoTheBlackHole** [stashapp/stash#3288]
**Is your feature request related to a problem? Please describe.**
When I delete something from stash, I neither want to delete it permanently nor only from stashdb. I need it to be moved to a trash folder first and then remove it from stashdb.  
This is how many apps on Linux work. They use trash-cli to delete as a Recycle Bin. 

**Describe the solution you'd like**
Allow satshapp to use a trash-cli or something else which better supports trash management.  
for example:  https://pkg.go.dev/github.com/rkoesters/xdg/trash

**Comment by TamalesEnMiPinata** [comment]
It is desirable to have a Recycle Bin path configurable in stash, similar to other paths:
![image](https://github.com/user-attachments/assets/4f133fd5-cfae-40fb-899c-249a8d7e29d2)

If the Recycle Bin path is set (perhaps there would be an additional option to "Enable Recycle Bin"), when a user chooses to delete a scene the normal database updates would occur and as far as Stash is concerned the scene no longer exists. However, in the file system the scene and corresponding media file(s) and generated data would be moved to the Recycle Bin path while preserving directory structure.

**Comment by WithoutPants** [comment]
I fairly simple implementation would be to add an option to enter a command line to use when deleting media files. This would allow the option to move the files instead via `mv {} /trash/` for example, or to use trash-cli via `trash-cli {}`

**Comment by Gykes** [comment]
@WithoutPants Would you like this to apply over all all OS's? Could do ` trash-cli {}` or `mv {} /trash/` for linux and `mv {} ~/.Trash/{}` for mac.

For windows it's a little more complicated as I think we need to include powershell or something into the loop.

**Comment by smith113-p** [comment]
Windows has SHFileOperation for that purpose and Google tells me it is possible to call that from go.

**Comment by Gykes** [comment]
Yea, that looks right based off some googling. 

Do we want the path to be configurable based on OS or hardcode it and just make it an enable/disable option in settings?

**Comment by TamalesEnMiPinata** [comment]
> Do we want the path to be configurable based on OS or hardcode it and just make it an enable/disable option in settings?

I'm not sure I know the right language to answer that but I would envision the ability in my docker configuration to be able to specify where the "trash" path is. For example, I use the Unraid RecycleBin plugin which puts a .RecycleBin folder in the root of every share, so "/mnt/user/stashData/.RecycleBin".

It sounds like you guys have a different idea on how to actually implement this, but my initial thought was that I would just have a mapping:

Container Path: /trash
Host Path: /mnt/user/stashData/.RecycleBin

Obviously not everybody is using Docker but I assume there is some equivalent to running Stash on windows to Docker mount paths.

Just my 2 cents.

**Comment by Gykes** [comment]
> > Do we want the path to be configurable based on OS or hardcode it and just make it an enable/disable option in settings?
> 
> I'm not sure I know the right language to answer that but I would envision the ability in my docker configuration to be able to specify where the "trash" path is. For example, I use the Unraid RecycleBin plugin which puts a .RecycleBin folder in the root of every share, so "/mnt/user/stashData/.RecycleBin".
> 
> It sounds like you guys have a different idea on how to actually implement this, but my initial thought was that I would just have a mapping:
> 
> Container Path: /trash Host Path: /mnt/user/stashData/.RecycleBin
> 
> Obviously not everybody is using Docker but I assume there is some equivalent to running Stash on windows to Docker mount paths.
> 
> Just my 2 cents.

This is my plan :)

**Comment by WithoutPants** [comment]
I think there's a few different options we could explore. One pretty simple way is a configurable trash location as mentioned:
add `trash` location to `System -> Application Paths`; if non-empty, move media files here instead of deleting.

My idea [above](https://github.com/stashapp/stash/issues/3288#issuecomment-3447903117) was to add a configurable `trash` command, so that uses can enter a command to use when deleting. Very low-level, but potentially powerful and highly configurable. The implementation is left for the user to figure out basically, and the format of the command will be based on the host OS.

The other option is to use platform-specific methods to use the applicable API to (optionally) move items to trash. This will work nicely for some OS's and requires the least configuration, but I'm not overly keen on adding platform specific code in.

I think a trash location is the simplest first step.

**Comment by Gykes** [comment]
I PR'd a rendition of this. 

Essentially it should be considered a decent first step. Currently you can just add a path in your Stash settings. If the path is populated then all deleted files should go to that path. If the path does not exist or the setting is empty then the usual permanent deletion will persist. 

There is some logging to let users if the path does not exist and that the file was permanently deleted. Please make sure your path is valid before yee hawing all of your files to trash.

**Comment by smith113-p** [comment]
There's also https://github.com/hymkor/trash-go

(I don't personally have an opinion on which way to go)

**Comment by Gykes** [comment]
> There's also https://github.com/hymkor/trash-go
> 
> (I don't personally have an opinion on which way to go)

From my 5 minutes of looking at that it seemed like it only allows users to send files to the default trash location for each OS. My way allows the user to specify a location. The location the user wants to define may not be the default "Trash" but just a folder that they want things to go in case of a delete. 

Also, Stash already had like 95% of the infrastructure to handle moving a file to a trash folder. This way seemed a lot simpler and gave the user more control so it was a win win in my book

I could be totally wrong on my assumption and i'm happy to be corrected.

**Comment by smith113-p** [comment]
Yeah, that's true, the main benefit would be handling windows correctly and automatically finding the default trashcan path on other OSes, but you're right (I think) that it does not let you customize the trash path

**Comment by Gykes** [comment]
I could hard wire some defaults in. The issue with that is not everyone wants this option enabled, I know I don't, and will find it annoying to have to essentially delete files twice.

Realistically, in my opinion anyway, a trash can is just any folder in any location that you decide is where your garbage goes. It doesn't need to specifically be the recycle bin or anything like that.

**Comment by smith113-p** [comment]
I disagree with that definition of a trash can. IMO a core feature is easy undeletion, which requires also storing the original location of the file and UI for that.

(Looks like that is another benefit of that library, it does write that metadata)

**Comment by Gykes** [comment]
Fair enough, if you use the correct path then all of those features should be available to you. 

For people who just want a dump folder, they can also do it.

Edit: I'm not actually sure if moving to the recycle bin will store previous save metadata, would need to do testing with it.

**Comment by smith113-p** [comment]
> Fair enough, if you use the correct path then all of those features should be available to you.
> 
> For people who just want a dump folder, they can also do it.
> 
> Edit: I'm not actually sure if moving to the recycle bin will store previous save metadata, would need to do testing with it.

Yes, I am almost certain that it won't. The library I linked writes a trashinfo file for that purpose: https://github.com/hymkor/trash-go/blob/master/internal/freedesktop/main.go#L112 (see also https://specifications.freedesktop.org/trash/latest/#id7)

On Windows, SHFileOperation handles the metadata, but a regular os.Rename won't.

But I don't know if the original requestor wanted the simpler version you are proposing or the standard trash bin

**Comment by Gykes** [comment]
> > Fair enough, if you use the correct path then all of those features should be available to you.
> > 
> > For people who just want a dump folder, they can also do it.
> > 
> > Edit: I'm not actually sure if moving to the recycle bin will store previous save metadata, would need to do testing with it.
> 
> Yes, I am almost certain that it won't. The library I linked writes a trashinfo file for that purpose: https://github.com/hymkor/trash-go/blob/master/internal/freedesktop/main.go#L112 (see also https://specifications.freedesktop.org/trash/latest/#id7)
> 
> On Windows, SHFileOperation handles the metadata, but a regular os.Rename won't.
> 
> But I don't know if the original requestor wanted the simpler version you are proposing or the standard trash bin

I have no issue going this route just ultimately it will be up to withoutpants. He is the main dev and I am but a lowly worker. 

For @WithoutPants,

If we go my route it gives users configurable options for where to set trash but you lose some options that trash has.

If you go the route of the go package it seems like we lose configurable options but users gain the abilities of a real "trash"

Thoughts?
